File: typescript/src/problems/s1588_sum_of_all_odd_length_subarrays.ts
/**
 * @link https://leetcode.com/problems/sum-of-all-odd-length-subarrays/
 * @param arr
 */
export function sumOddLengthSubarrays(arr: number[]): number {
  const prefixSum = new Array(arr.length + 1).fill(0);
  for (let i = 0; i < arr.length; i++) {
    prefixSum[i + 1] = prefixSum[i] + arr[i];
  }
  let sum = 0;
  for (let i = 0; i < arr.length; i++) {
    for (let j = 1; j + i <= arr.length; j += 2) {
      sum += prefixSum[j + i] - prefixSum[i];
    }
  }
  return sum;
}

File: rust/src/problems/s1588_sum_of_all_odd_length_subarrays.rs
struct Solution;

impl Solution {
    pub fn sum_odd_length_subarrays(arr: Vec<i32>) -> i32 {
        let mut prefix_sum = vec![0; arr.len() + 1];
        for i in 0..arr.len() {
            prefix_sum[i + 1] = prefix_sum[i] + arr[i];
        }
        let mut sum = 0;
        for i in 0..arr.len() {
            let mut j = 1;
            while j + i <= arr.len() {
                sum += prefix_sum[j + i] - prefix_sum[i];
                j += 2;
            }
        }
        sum
    }
}

struct Example {
    input: Vec<i32>,
    output: i32,
}

#[test]
pub fn test() {
    let examples = vec![
        Example {
            input: vec![1, 4, 2, 5, 3],
            output: 58,
        },
        Example {
            input: vec![1, 2],
            output: 3,
        },
        Example {
            input: vec![10, 11, 12],
            output: 66,
        },
    ];
    for example in examples {
        assert_eq!(Solution::sum_odd_length_subarrays(example.input), example.output);
    }
}
